pyomo/contrib/community_detection/detection.py

```diff
@@ -168,7 +170,225 @@ def detect_communities(model, type_of_community_map='constraint', with_objective
+            'constraint' draws a graph with constraint nodes,
+            'variable' draws a graph with variable nodes,
+            'bipartite' draws a bipartite graph (with both constraint and variable nodes).
+        with_objective: bool, optional
+            a Boolean argument that specifies whether or not the objective function is included in the graph; the
+            default is True
+        weighted_graph: bool, optional
+            (this argument is only used if no community_map is provided)
+            a Boolean argument that specifies whether a community map is created based on a weighted graph or an
+            unweighted graph; the default is True (type_of_community_map='bipartite' creates an unweighted graph
+            regardless of this parameter)
+        random_seed: int, optional
+            (this argument is only used if no community_map is provided)
+            an integer that is used as the random seed for the (heuristic) Louvain community detection
+        pos: dict, optional
+            a dictionary that maps node keys to their positions on the illustration
+
+        Returns
+        -------
+        fig: matplotlib figure
+            the figure for the model graph drawing - can be illustrated by calling 'plt.show()'
+        pos: dict
+            a dictionary that maps node keys to their positions on the illustration - can be used to create consistent
+            layouts for graphs of a given model
+        """
+
+        # Check that all arguments are of the correct type
+
+        assert type_of_graph in ('bipartite', 'constraint', 'variable'), \
+            "Invalid graph type specified: 'type_of_graph=%s' - Valid values: " \
+            "'bipartite', 'constraint', 'variable'" % type_of_graph
+
+        assert type(with_objective) == bool, "Invalid value for with_objective: 'with_objective=%s' - with_objective " \
+                                             "must be a Boolean" % with_objective
+
+        assert type(weighted_graph) == bool, "Invalid value for weighted_graph: 'weighted_graph=%s' - weighted_graph " \
+                                             "must be a Boolean" % weighted_graph
+
+        assert random_seed is None or (type(random_seed) == int and random_seed >= 0), \
+            "Invalid value for random_seed: 'random_seed=%s' - random_seed must be a non-negative integer" % random_seed
+
+        # No assert statement for pos; the NetworkX function can handle issues with the pos argument
+
+        # Use the generate_model_graph function to create a NetworkX graph of the given model (along with
+        # number_component_map and constraint_variable_map, which will be used to help with drawing the graph)
+        model_graph, number_component_map, constraint_variable_map = generate_model_graph(
+            self.model, type_of_graph=type_of_graph, with_objective=with_objective, weighted_graph=False)
```

Comment:
Should the networkx representation be an attribute of the CommunityMap object?

Reply:
The networkX graph generated in those lines may be a little different from the one used to create the CommunityMap object (based on my response above) - however, I can add the networkX graph as an attribute and then also add a check before these lines to avoid creating the same networkX graph as we already made to create the CommunityMap object